File: crates/tcode-tui/src/app/replay.rs
//! Rebuilding the transcript from a stored ledger.
//!
//! Replay must produce exactly what the live path produced, so it goes
//! through the same `bake_*` entry points rather than formatting anything
//! itself — batch grouping, blank-line separators and diff bodies have all
//! drifted apart here before.
//!
//! Touches: transcript, renderers, md, pending_tool, pending_batch,
//! task_runs, space_before_response.

use super::*;

/// A tool call recovered from the ledger during replay, with the batch it
/// was executed in (asked of core, never re-derived here).
pub(super) struct ReplayCall {
    pub(super) id: String,
    pub(super) name: String,
    pub(super) input: serde_json::Value,
    pub(super) batch: Option<ReplayBatch>,
}

pub(super) struct ReplayBatch {
    /// The `● label` header, carried by the batch's first call only.
    pub(super) header: Option<String>,
    /// Several tools in one batch: tag each item with its tool name.
    pub(super) mixed: bool,
}

impl App {
    /// Replay a resumed conversation into the scrollback so the user
    /// sees where they left off.
    pub(super) fn bake_transcript(&mut self) {
        // Moved out for the walk so the bake helpers below can take `&mut
        // self`; restored before returning.
        let Some(session) = self.session.take() else {
            return;
        };
        if !session.ledger.is_empty() {
            self.replay_ledger(&session);
        }
        self.session = Some(session);
    }

    pub(super) fn replay_ledger(&mut self, session: &Session) {
        let mut lines: Vec<Line<'static>> = Vec::new();
        let mut resumed_progress: Option<serde_json::Value> = None;
        let mut calls: HashMap<String, ReplayCall> = HashMap::new();
        let mut space_before_assistant_text = false;
        for (entry_index, entry) in session.ledger.entries().iter().enumerate() {
            match entry {
                tcode_core::Entry::User(blocks) => {
                    space_before_assistant_text = false;
                    // User echoes are their own entry-tagged blocks so
                    // rewind can jump to and truncate from them.
                    self.transcript.push(std::mem::take(&mut lines));
                    let mut echo: Vec<Line<'static>> = vec![Line::default()];
                    for b in blocks {
                        match b {
                            ContentBlock::Text { text } if !text.starts_with("<tcode-status>") => {
                                echo.extend(quote_lines(None, text));
                            }
                            ContentBlock::Image { .. } => {
                                echo.push(quote_attachment_line("[image]"));
                            }
                            _ => {}
                        }
                    }
                    // Keep a breathing row between a highlighted human
                    // message and the following assistant/tool activity.
                    echo.push(Line::default());
                    self.transcript.push_tagged(echo, entry_index);
                }
                tcode_core::Entry::Assistant(blocks) => {
                    let mut group: Vec<(String, String, serde_json::Value)> = Vec::new();
                    for b in blocks {
                        match b {
                            ContentBlock::Thinking { thinking, .. } => {
                                // Its own foldable block, like live streaming.
                                // The duration is not recorded, so the head
                                // states only the size.
                                self.transcript.push(std::mem::take(&mut lines));
                                let title =
                                    format!("reasoning (~{} tok)", thinking.chars().count() / 3);
                                self.bake_thinking(&title, thinking);
                            }
                            ContentBlock::Text { text } => {
                                if space_before_assistant_text {
                                    self.transcript.push(std::mem::take(&mut lines));
                                    self.bake(vec![Line::default()]);
                                    space_before_assistant_text = false;
                                }
                                self.transcript.push(std::mem::take(&mut lines));
                                self.transcript
                                    .push_markdown(self.md.parse(text).with_trailing_blank());
                            }
                            ContentBlock::ToolUse { id, name, input } => {
                                // Defer the header to the matching ToolResults
                                // entry so each call renders directly above its
                                // own result, not all headers then all results.
                                group.push((id.clone(), name.clone(), input.clone()));
                                if matches!(self.renderers.get(name).route(), CallRoute::Progress) {
                                    resumed_progress = Some(input.clone());
                                }
                            }
                            _ => {}
                        }
                    }
                    // Ask core whether these calls ran as a batch: the loop
                    // that made that call is the only place that knows.
                    let label = self.agent.batch_display_label(session, &group);
                    let mixed = group
                        .iter()
                        .map(|(_, name, _)| name.as_str())
                        .collect::<HashSet<_>>()
                        .len()
                        > 1;
                    for (index, (id, name, input)) in group.into_iter().enumerate() {
                        let batch = label.as_ref().map(|label| ReplayBatch {
                            header: (index == 0).then(|| label.clone()),
                            mixed,
                        });
                        calls.insert(
                            id.clone(),
                            ReplayCall {
                                id,
                                name,
                                input,
                                batch,
                            },
                        );
                    }
                }
                tcode_core::Entry::IncompleteAssistant { text, error } => {
                    self.transcript.push(std::mem::take(&mut lines));
                    self.transcript
                        .push_markdown(self.md.parse(text).with_trailing_blank());
                    lines.push(Line::styled(
                        format!(
                            "↻ stream failed: {error} — incomplete response retained; not sent back to model"
                        ),
                        theme::error_highlight(),
                    ));
                    lines.push(Line::default());
                }
                tcode_core::Entry::Summary(summary) => {
                    // Flush the prose above so the divider bakes as its own
                    // foldable block, exactly as the live path lays it out.
                    self.transcript.push(std::mem::take(&mut lines));
                    let summary = summary.clone();
                    self.bake_compacted(&summary);
                }
                tcode_core::Entry::ImportedTool {
                    name,
                    input,
                    content,
                } => {
                    // These names come from external logs (Codex / Claude
                    // Code) and are never in the render registry; imported
                    // rendering stays hardcoded by design.
                    if name.contains("apply_patch") {
                        lines.push(Line::from(vec![
                            Span::styled("● ", theme::accent()),
                            Span::styled(
                                format!("{name} (imported historical change)"),
                                theme::bold(),
                            ),
                        ]));
                        lines.extend(diff::render_unified_patch(content));
                    } else if name == "output" {
                        for (index, line) in content.lines().enumerate() {
                            let prefix = if index == 0 { "  ⎿ " } else { "    " };
                            lines.push(Line::styled(format!("{prefix}{line}"), theme::dim()));
                        }
                    } else {
                        let summary = if input.is_null() {
                            name.clone()
                        } else {
                            self.display_summary(&tcode_core::agent::summarize_call(name, input))
                        };
                        lines.push(Line::from(vec![
                            Span::styled("● ", theme::accent()),
                            Span::styled(summary, theme::bold()),
                        ]));
                        if !content.is_empty() {
                            self.transcript.push(std::mem::take(&mut lines));
                            self.transcript
                                .push_markdown(self.md.parse(content).with_trailing_blank());
                        }
                    }
                    lines.push(Line::default());
                }
                tcode_core::Entry::ToolResults(blocks) => {
                    for block in blocks {
                        let ContentBlock::ToolResult {
                            tool_use_id,
                            content,
                            is_error,
                            ..
                        } = block
                        else {
                            continue;
                        };
                        let call = calls.get(tool_use_id);
                        let name = call.map(|c| c.name.as_str()).unwrap_or("tool");
                        // Plan/Silent calls render via the plan pane or their
                        // approval record, exactly like the live path.
                        if !matches!(self.renderers.get(name).route(), CallRoute::Transcript) {
                            continue;
                        }
                        // Flush the prose above so this call bakes as its own
                        // block, exactly as the live path lays it out.
                        self.transcript.push(std::mem::take(&mut lines));
                        // Bake this call's header (+ diff / command block)
                        // right above its own result.
                        let run_id = call.and_then(|call| {
                            self.task_runs
                                .iter()
                                .find(|run| run.parent_call == call.id)
                                .map(|run| run.id.clone())
                        });
                        let blocks_before = self.transcript.block_count();
                        let record = match call {
                            Some(call) => match &call.batch {
                                Some(batch) => {
                                    if let Some(label) = &batch.header {
                                        let header = self.batch_header_lines(label);
                                        self.bake(header);
                                    }
                                    CallRecord::Batch(self.batch_item_lines(
                                        name,
                                        &call.input,
                                        batch.mixed,
                                    ))
                                }
                                None => match self.bake_call_start(name, &call.input) {
                                    Some(index) => CallRecord::HeaderBlock(index),
                                    None => CallRecord::Baked,
                                },
                            },
                            None => CallRecord::Baked,
                        };
                        let preview = result_preview(content);
                        let report = run_id.as_ref().map(|_| task_result_text(content));
                        self.bake_call_result(
                            name,
                            call.map(|c| &c.input),
                            &preview,
                            report.as_deref().unwrap_or(content),
                            *is_error,
                            record,
                        );
                        if let Some(run) =
                            run_id.filter(|_| self.transcript.block_count() > blocks_before)
                        {
                            if let Some(index) = self.transcript.last_block_index() {
                                self.transcript.link_task_run(index, run.clone());
                                if let Some(entry) =
                                    self.task_runs.iter_mut().find(|entry| entry.id == run)
                                {
                                    entry.block = Some(index);
                                }
                            }
                        }
                        space_before_assistant_text = true;
                    }
                }
                tcode_core::Entry::UserNote { text, .. } => {
                    // Approval annotations and `ask_user` answers both retain
                    // the person's original wording on resume. Live questions
                    // have a richer Q&A record, but that UI-only shape is not
                    // persisted in the ledger.
                    self.bake_user_note(text);
                }
                tcode_core::Entry::Note(text) => {
                    lines.push(Line::default());
                    lines.extend(quote_lines(Some(NOTE_LABEL), text));
                    lines.push(Line::default());
                }
            }
        }
        lines.push(Line::styled("── resumed ──", theme::dim()));
        if let Some(progress) = resumed_progress {
            self.update_progress(&progress);
        }
        self.bake(lines);
    }
}

File: crates/tcode-app/ui/src/replay.ts
import { applyEvent, type Block } from "./blocks";
import { applyFileEvent, type TouchedFile } from "./files";
import type { AgentEvent, LedgerEntry } from "./types";

/** The persisted ledger is authoritative after a restart; live events no longer
 * exist. Rebuild the same display model that the event stream normally feeds. */
export function replayLedger(history: LedgerEntry[]): { blocks: Block[]; files: TouchedFile[] } {
  let blocks: Block[] = [];
  let files: TouchedFile[] = [];
  const feed = (event: AgentEvent) => {
    blocks = applyEvent(blocks, event);
    files = applyFileEvent(files, event);
  };

  for (const entry of history) {
    switch (entry.kind) {
      case "user": {
        const content = blocksOf(entry.data);
        const text = content
          .filter(isText)
          .map((block) => block.text)
          .filter((text) => !text.startsWith("<tcode-status>"))
          .join("\n");
        const images = content.filter(isImage).map((block) => `data:${block.media_type};base64,${block.data}`);
        if (text || images.length > 0) {
          blocks = [...blocks, { kind: "user", text, images: images.length > 0 ? images : undefined }];
        }
        break;
      }
      case "assistant":
        for (const block of blocksOf(entry.data)) {
          if (isText(block)) feed({ type: "TextDelta", data: block.text });
          if (isThinking(block)) feed({ type: "ThinkingDelta", data: block.thinking });
          if (isToolUse(block)) {
            feed({
              type: "ToolStart",
              data: { call_id: block.id, name: block.name, summary: block.name, input: block.input },
            });
          }
        }
        break;
      case "tool_results":
        for (const block of blocksOf(entry.data)) {
          if (isToolResult(block)) {
            feed({
              type: "ToolEnd",
              data: {
                call_id: block.tool_use_id,
                name: "tool",
                preview: preview(block.content),
                content: block.content,
                is_error: block.is_error,
              },
            });
          }
        }
        break;
      case "note":
        feed({ type: "Note", data: stringOf(entry.data) });
        break;
      case "user_note": {
        const note = recordOf(entry.data);
        feed({
          type: "UserNote",
          data: { text: typeof note.text === "string" ? note.text : "", answer: note.answer === true },
        });
        break;
      }
      case "summary":
        feed({ type: "Compacted", data: stringOf(entry.data) });
        break;
      case "incomplete_assistant": {
        const incomplete = recordOf(entry.data);
        if (typeof incomplete.text === "string") feed({ type: "TextDelta", data: incomplete.text });
        if (typeof incomplete.error === "string") {
          blocks = [...blocks, { kind: "error", text: `stream failed: ${incomplete.error}` }];
        }
        break;
      }
      case "imported_tool": {
        const imported = recordOf(entry.data);
        const name = typeof imported.name === "string" ? imported.name : "imported tool";
        const callId = `imported-${blocks.length}`;
        feed({ type: "ToolStart", data: { call_id: callId, name, summary: name, input: imported.input } });
        feed({
          type: "ToolEnd",
          data: {
            call_id: callId,
            name,
            preview: preview(typeof imported.content === "string" ? imported.content : ""),
            content: typeof imported.content === "string" ? imported.content : "",
            is_error: false,
          },
        });
        break;
      }
      // Instructions are filtered by the backend. Unknown future entries stay
      // invisible rather than turning persisted data into a broken transcript.
      default:
        break;
    }
  }

  return { blocks, files };
}

type PersistedBlock = Record<string, unknown> & { type?: string };

function blocksOf(value: unknown): PersistedBlock[] {
  return Array.isArray(value) ? value.filter((block): block is PersistedBlock => typeof block === "object" && block !== null) : [];
}

function recordOf(value: unknown): Record<string, unknown> {
  return typeof value === "object" && value !== null ? value as Record<string, unknown> : {};
}

function stringOf(value: unknown): string {
  return typeof value === "string" ? value : "";
}

function isText(block: PersistedBlock): block is PersistedBlock & { text: string } {
  return block.type === "text" && typeof block.text === "string";
}

function isThinking(block: PersistedBlock): block is PersistedBlock & { thinking: string } {
  return block.type === "thinking" && typeof block.thinking === "string";
}

function isImage(block: PersistedBlock): block is PersistedBlock & { media_type: string; data: string } {
  return block.type === "image" && typeof block.media_type === "string" && typeof block.data === "string";
}

function isToolUse(block: PersistedBlock): block is PersistedBlock & { id: string; name: string; input: unknown } {
  return block.type === "tool_use" && typeof block.id === "string" && typeof block.name === "string";
}

function isToolResult(block: PersistedBlock): block is PersistedBlock & { tool_use_id: string; content: string; is_error: boolean } {
  return block.type === "tool_result" && typeof block.tool_use_id === "string" && typeof block.content === "string" && typeof block.is_error === "boolean";
}

function preview(content: string): string {
  return content.split("\n", 1)[0] || "done";
}
